14.py

```diff
@@ -0,0 +1,160 @@
+                print("Cannot pour more water than the tank's capacity.")
+
+    # WYLEWAMY WODĘ
+    def pour_out_water(self, volume):
+        if volume > 0:
+            if volume <= self.water_level:
+                self.water_level -= volume
+                self.record_operation("pour_out_water", volume, success=True)
+            else:
+                self.record_operation("pour_out_water", volume, success=False)
+                print("Cannot pour out more water than is available in the tank.")
+
+    # PRZELEWAMY WODĘ DO INNEGO ZBIORNIKA
+    def transfer_water(self, from_tank, volume):
+        # I. condition - poziom wody do przelania musi być większy od zera
+        if volume > 0:
+
+            # II. condition - poziom wody w tanku, z którego będzie brana woda,
+            # musi być wyższy albo równy ilości wody do transferu
+            if from_tank.water_level >= volume:
+
+                # III. condition - poziom wody po przelaniu musi mieścić się w capacity tanku,
+                # do którego ma być przelana woda
+                if self.water_level + volume <= self.capacity:
+                    # Operacje zmiany wartości wody w zbiornikach
+                    from_tank.water_level -= volume
+                    self.water_level += volume
+
+                    # Zapisz operację w historii obu zbiorników
+                    self.record_operation("transfer_water_out", volume, success=True, from_tank=from_tank, to_tank=self)
+                    from_tank.record_operation("transfer_water_in", volume, success=True, from_tank=from_tank,
+                                               to_tank=self)
+
+                else:
+                    self.record_operation("transfer_water_out", volume, success=False, from_tank=from_tank)
+                    print("Cannot transfer more water than the receiving tank's capacity.")
+            else:
+                self.record_operation("transfer_water_out", volume, success=False, from_tank=from_tank)
+                print("Cannot transfer more water than is available in the source tank.")
+        else:
+            print("Water volume must be greater than zero!")
+
+    def check_state(self):
+        current_water_level = 0
+
+        for operation in self.operations_history:
+            if operation[3]:  # Sprawdź, czy operacja była udana
+                if operation[4] == "pour_water" and operation[0] == self.name:
+                    current_water_level += operation[1]
+                elif operation[4] == "pour_out_water" and operation[0] == self.name:
+                    current_water_level -= operation[1]
+                elif operation[4] == "transfer_water_out":
+                    # Źródło wody (from_tank) odejmuje wodę
+                    current_water_level += operation[1]
+                elif operation[4] == "transfer_water_in":
+                    # Docelowy zbiornik dodaje wodę
+                    current_water_level -= operation[1]
```

Comment:
to jest sterowanie logiką za pomocą intrrukcji warunkowych
alternatywa - sterowanie logiką za pomocą mapowania
```py
operation_name_to_volume = {
   "transfer_water_out": lambda volume: volume
   "transfer_water_in": lambda volume: -volume
}
current_water_level += operation_name_to_volume[operation[4]](operation[1])
```